### ISSUE TITLE
dpo training with Qwen2.5-0.5B+1000 seq length cost 17GB GPU mem

### ISSUE THREAD

**Issue by 32234** [huggingface/trl#5200]
### Reproduction

using trl==0.12.0

```
def f(example):
      example['prompt'] = 'you are person'*500
      example['chosen'] = '1'
      example['rejected'] = '0'
      return example



trainset = dataset.map(f,remove_columns)

from transformers import AutoModelForCausalLM, BitsAndBytesConfig
import torch

model_name = './Qwen2.5-0.5B'

bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype="float16",  # Compute dtype
    bnb_4bit_use_double_quant=True,
)




model = AutoModelForCausalLM.from_pretrained(
    model_name,
    quantization_config=bnb_config,
    attn_implementation="flash_attention_2",
    device_map="auto"
)

model.config.use_cache = False
model.config.pretraining_tp = 1


tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True)
** tokenizer.pad_token = "<PAD>"
tokenizer.padding_side = "left"


from peft import LoraConfig, prepare_model_for_kbit_training, get_peft_model


peft_config = LoraConfig(
    lora_alpha=32,  # LoRA Scaling
    lora_dropout=0.1,  # Dropout for LoRA Layers
    r=64,  # Rank
    bias="none",
    task_type="CAUSAL_LM",
    target_modules=  # Layers to target
    #['k_proj', 'v_proj',  'q_proj']
    ['k_proj', 'gate_proj', 'v_proj', 'up_proj', 'q_proj', 'o_proj', 'down_proj']
)


model = prepare_model_for_kbit_training(model)
model = get_peft_model(model, peft_config)


from trl import DPOConfig

output_dir = "./results"

training_arguments = DPOConfig(
    output_dir=output_dir,
    per_device_train_batch_size=2,
    gradient_accumulation_steps=4,
    optim="paged_adamw_32bit",
    learning_rate=2e-5,
    lr_scheduler_type="cosine",
    max_steps=100,
    logging_steps=10000,
    fp16=True,
    gradient_checkpointing=True,
    warmup_ratio=0.1,
  ** precompute_ref_log_probs=True
)


from trl import DPOTrainer

from transformers import TrainerCallback

class MemoryMonitorCallback(TrainerCallback):
    def on_step_end(self, args, state, control, **kwargs):
        if state.global_step % 10 == 0:
            peak = torch.cuda.max_memory_allocated() / 1024**3
            print(f"Peak in this step: {peak:.2f} GB")
            allocated = torch.cuda.memory_allocated() / 1024**3
            reserved = torch.cuda.memory_reserved() / 1024**3
            torch.cuda.reset_peak_memory_stats()
            print(f"\n[Step {state.global_step}] GPU Memory - Allocated: {allocated:.2f} GB | Reserved: {reserved:.2f} GB")

dpo_trainer = DPOTrainer(
    model,
    args=training_arguments,
    train_dataset=trainset,
    tokenizer=tokenizer,
    beta=0.1,
    max_prompt_length=2048,
    max_length=3000,
    callbacks=[MemoryMonitorCallback()]
)



dpo_trainer.train()

```

### then running it output:
Peak in this step: 15.14 GB
[Step 10] GPU Memory - Allocated: 0.76 GB | Reserved: 17.16 GB,  

I opened grad-checkpoint,  flashattention,   precompute_ref_log_probs, but still cost 17GB, could someone please tell  why and how it costs so much memory ??

### System Info

using trl==0.12.0  transformers 4.57 python==3.12

**Comment by qgallouedec** [comment]
please format your code using code blocks. Also please upgrade trl to its latest version

**Comment by 32234** [comment]
> please format your code using code blocks. Also please upgrade trl to its latest version

thanks, I'll try and report the result here

**Comment by albertvillanova** [comment]
Hey @32234, sorry for the slow reply — catching up on the backlog. 

The 17 GB number you observed is largely expected behavior, not a bug: DPO does a forward+backward over both chosen and rejected sequences (so effectively 4×3000 tokens per micro-batch here), and at `max_length=3000` activations dominate even with gradient checkpointing and a 4-bit base model. `precompute_ref_log_probs=True` only removes the *reference* model's forward — the policy model still needs full activations for backprop. You can reduce memory by lowering `max_length`/`max_prompt_length`, dropping `per_device_train_batch_size` to 1 (compensating with `gradient_accumulation_steps`), or removing unneeded LoRA target modules. 

Closing as not-a-bug; v0.12.0 is also quite old at this point — if you upgrade to the latest TRL and still see something that looks pathological (e.g. memory not scaling as expected with `max_length`), please open a fresh issue with the numbers and we'll take another look.

**Comment by 32234** [comment]
> Hey [@32234](https://github.com/32234), sorry for the slow reply — catching up on the backlog.
> 
> The 17 GB number you observed is largely expected behavior, not a bug: DPO does a forward+backward over both chosen and rejected sequences (so effectively 4×3000 tokens per micro-batch here), and at `max_length=3000` activations dominate even with gradient checkpointing and a 4-bit base model. `precompute_ref_log_probs=True` only removes the _reference_ model's forward — the policy model still needs full activations for backprop. You can reduce memory by lowering `max_length`/`max_prompt_length`, dropping `per_device_train_batch_size` to 1 (compensating with `gradient_accumulation_steps`), or removing unneeded LoRA target modules.
> 
> Closing as not-a-bug; v0.12.0 is also quite old at this point — if you upgrade to the latest TRL and still see something that looks pathological (e.g. memory not scaling as expected with `max_length`), please open a fresh issue with the numbers and we'll take another look.

thanks, I finally understand that it costs 4x activations memory with both chosen and reject answer when batch_size = 2, then it makes sense to me. Maybe the old version of trl did less on gradient_checkpointing
